### ISSUE TITLE
hinted handoff may end up in a live-lock due to write timeouts

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#7886]
*Installation details*
HEAD: 9eed26ca3d82dadef4cf7c4034cc39bfe5d94c70
Cluster size: 2+

*Description*
Hinted handoff sender may find itself in a live-lock situation if hints sending always fails due to a timeout.

There are a few things that may contribute to this:
1) A fixed concurrency (10% of the total memory in in-flight buffers) of a sender. If this concurrency leads to the overload of the receiver and received hints begin to queue up this will result in a stale-mate situation when sender will re-send same hints over and over again and thereby not allow the receiver to recover.

2) Multiple hints senders sending to the same node (e.g. after some time the receiver node was DOWN).

*Proposed solution*
There are two things that can be done. They are not in conflict and may be implemented independently.

1) Introduce a parameter that defines a maximum hints sender concurrency (in bytes).
It will control the `resource_manager::_max_send_in_flight_memory` value.

This is an easy hack that will allow us manually "fix" cases when we hint this issue.

2) Introduce a concurrency exponential backoff in the hints sender (note that the HH concurrency is currently expressed in bytes):
 * Cut the allowed concurrency in half every time the hints sending fails due to a timeout. The concurrency can go as low as 1 byte).
 * Every time we are able to successfully send a hint out - increase the concurrency by the value of the successfully send hint buffer size. The concurrency can increase up to the `resource_manager::_max_send_in_flight_memory`.

**Comment by vladzcloudius** [comment]
@slivne @piodul @psarna FYI

**Comment by slivne** [comment]
Need to start with a reproducer showing this can happen and then provide a solution

**Comment by vladzcloudius** [comment]
> Need to start with a reproducer showing this can happen and then provide a solution

@slivne I agree in general. However this issue has been discovered by simply reading the code.
I think that the description of possible use cases makes it pretty obvious that it IS possible to get to this situation in the field.

As to actual reproducer for Eng to work with - I'll try to cook something but probably sometime next week. You, guys, can try to play with it yourself in the meantime. Things to play with are the size of the payload and write timeout config: e.g. we can create hints population (by dropping one node) for which we know their write latency (X) if there is no queuing (namely in a 1-pending state) and if we define a payload big enough it may be as big as we want. Then we can simply reduce write_timeout to be a little more than X  and bring the node back up. As a result hints should start timing out because of hints concurrency which would be greater than 1.

**Comment by tarzanek** [comment]
just my 2 cents (or likely a silly Q) without reading the code (but I will try to have a look)
- they use LOGGED batches - will this somehow impact writing of hints? (ala do they ALL have to succeed or bust? is the timeout in such case per hint or per total batch?)

**Comment by vladzcloudius** [comment]
> just my 2 cents (or likely a silly Q) without reading the code (but I will try to have a look)
> 
>     * they use LOGGED batches - will this somehow impact writing of hints? (ala do they ALL have to succeed or bust? is the timeout in such case per hint or per total batch?)

That's a very good question, @tarzanek . I'd guess that it would be per single write operation but would need to check the code (next week).

On top of that LOGGED batches can cause the issue as they see by its own since they have their own replay logic with a small difference - it has MUCH LESS back pressure mechanisms there unlike HH.

**Comment by vladzcloudius** [comment]
@slivne, @asias Has created a reproducer: https://github.com/scylladb/scylla/issues/8863#issuecomment-863130996

**Comment by piodul** [comment]
@vladzcloudius 

The proposal makes sense to me. However, even if we implement this I think that it will still be possible to starve hint replay: although hints use the same write timeouts as normal writes, they are handled with lower CPU priority. In case when CPU is a bottleneck, hints may have much larger latency than regular writes and may keep timing out even when very low concurrency is used.

Come to think about it, sending hints is a maintenance operation and doesn't have strict latency requirements - therefore they don't really need the timeout and retrying them is just a wasted effort. In fact, hints used to have a very long timeout (1h), but I sent a patch which reduced it - because of how mutation sending logic is implemented, the previous timeout could delay shutdown by an hour. In retrospect, this issue could have been solved differently; for example, MV updates also use a long timeout and have custom logic for cancelling them if the destination goes down.

However, if we bring back the long timeout, the point 2. of this proposal will no longer be applicable. As for the point 1., it is already implemented but still waiting for merge: https://github.com/scylladb/scylla/pull/8646.

So, I see two options:

1. Go with this proposal and deal with occasional livelocks. It will still be better than what we have now.
2. Implement configurable concurrency for hints and bring back the long timeout. We would no longer have livelocks, but the algorithm is not adaptive anymore and hints may overwhelm regular writes if the hint concurrency parameter is chosen badly.

If we implement proper isolation for hints (https://github.com/scylladb/scylla/issues/8624) then we should choose option 2. without hesitation, but this is a longer project on which I am slowly working on.

**Comment by vladzcloudius** [comment]
> In case when CPU is a bottleneck, hints may have much larger latency than regular writes and may keep timing out even when very low concurrency is used.

If concurrency backs off all the way to 1 then hints latency is going to be very close to the latency of regular writes - maintenance CPU class has 300 shares while regular queries CPU class has 1000.

If hints keep on timing out with the lowest concurrency then you DO want to stop sending them for some time.

Hints sending *is* going to have some CPU even if CPU is a bottleneck and there shouldn't be a live-lock because of that.

If hints keep on timing out even with the lowest concurrency it means that regular writes are timing out at the same time too. At this point everything is "live-locked" - hints or not hints.

Static hints concurrency is not going to provide an answer in a generic case since whatever value you set - if it's more than 1 it still may be "too much".

But using it as a "max hints concurrency" + my proposal is going to make sure that hints sending are not choking replicas and allow regular CQL operations to proceed as long as it's possible - and this is the most important thing.

On the other hand, remember that regular hints are a "best effort" and we don't really care if they are send out or not.

The only hints that we care *at all* about are MV hints. 
But as I mentioned above - I don't see how (any) hints sending is going to get stuck if my proposal gets implemented. What MAY happen is that in case of an overload MV hints may start piling up and increase an "MV backlog" which will trigger MV flow control (FC) that is going to delay base table writes - and this is EXACTLY what you want in this case to reduce the overload.

About bringing long HH timeouts back - I don't see how this is going to improve anything. Having a "never ending" context can only bring trouble.
